ray_tile_and_stitch_util.py

```diff
@@ -0,0 +1,341 @@
+    # GT(3) y-coordinate of the upper-left corner of the upper-left pixel.
+    # GT(4) column rotation (typically zero).
+    # GT(5) n-s pixel resolution / pixel height (negative value for a north-up image).
+
+    # ---------------------- crop image from the water mask----------------------
+    # dot product of the mask and the orignal data before breaking it for processing
+    # Also band 2 3 and 4 are taken because the 4 bands cannot be processed by the NN learingin algo
+    # Need to make sure that the training bands are the same as the bands used for inferencing.
+    #
+    final_array_2 = input_image_gtif.GetRasterBand(2).ReadAsArray()
+    final_array_3 = input_image_gtif.GetRasterBand(3).ReadAsArray()
+    final_array_4 = input_image_gtif.GetRasterBand(4).ReadAsArray()
+
+    final_array_2 = np.multiply(final_array_2, mask)
+    final_array_3 = np.multiply(final_array_3, mask)
+    final_array_4 = np.multiply(final_array_4, mask)
+
+    # ulx, uly is the upper left corner
+    ulx, x_resolution, _, uly, _, y_resolution = input_image_gtif.GetGeoTransform()
+
+    # ---------------------- Divide image (tile) ----------------------
+    overlap_rate = 0.2
+    block_size = config.CROP_SIZE
+    ysize = input_image_gtif.RasterYSize
+    xsize = input_image_gtif.RasterXSize
+
+    # Close the file.
+    input_image_gtif = None
+    vfs.close_virtual_file()
+
+    tile_count = 0
+
+    y_list = range(0, ysize, int(block_size * (1 - overlap_rate)))
+    x_list = range(0, xsize, int(block_size * (1 - overlap_rate)))
+
+    # ---------------------- Find each Upper left (x,y) for each divided images ----------------------
+    #  ***-----------------
+    #  ***
+    #  ***
+    #  |
+    #  |
+    #
+    tiles = []
+    for id_i, i in enumerate(y_list):
+        # don't want moving window to be larger than row size of input raster
+        if i + block_size < ysize:
+            rows = block_size
+        else:
+            rows = ysize - i
+
+        # read col
+        for id_j, j in enumerate(x_list):
+            if j + block_size < xsize:
+                cols = block_size
+            else:
+                cols = xsize - j
+            # get block out of the whole raster
+            # todo check the array values is similar as ReadAsArray()
+            band_1_array = final_array_4[i: i + rows, j: j + cols]
+            band_2_array = final_array_2[i: i + rows, j: j + cols]
+            band_3_array = final_array_3[i: i + rows, j: j + cols]
+
+            # filter out black image
+            if (
+                band_3_array[0, 0] == 0
+                and band_3_array[0, -1] == 0
+                and band_3_array[-1, 0] == 0
+                and band_3_array[-1, -1] == 0
+            ):
+                continue
+
+            # stack three bands into one array
+            img = np.stack((band_1_array, band_2_array, band_3_array), axis=2)
+            cv2.normalize(img, img, 0, 255, cv2.NORM_MINMAX)
```

Comment:
I think that all this code could be simplified if we created the image stack above when we are reading the bands from the input geotif. If you did that, you could apply the mask dot product to all channels in one operations and getting the image would become something like 
```
img = final_array[i:i+rows, j:j+cols, :]
```

Reply:
tried doing that and ran into weird errors with numpy so I'm going to not implement this suggestion, it's also tangential to the ray migration. Thanks for the idea though :)